lib/axon/loop.ex

```diff
@@ -695,6 +721,37 @@ defmodule Axon.Loop do
     %Loop{loop | handlers: handle_fns}
   end
 
+  @doc """
+  Adds a handler function which logs the given message produced
+  by `message_fn` to the given IO device every `event` satisfying
+  `filter`.
+
+  In most cases, this is useful for inspecting the contents of
+  the loop state at intermediate stages. For example, the default
+  `trainer` loop factory attaches IO logging of epoch, batch, loss
+  and metrics.
+
+  It's also possible to log loop state to files by changing the
+  given IO device. By default, the IO device is `:stdio`.
+
+  `message_fn` should take the loop state and return a binary
+  representing the message to be written to the IO device.
+  """
+  def log(%Loop{} = loop, event, message_fn, device \\ :stdio, filter \\ :always)
+      when is_function(message_fn, 1) do
+    log_fn = fn %State{} = state ->
+      # TODO: Should we rescue here and terminate the loop
+      # with an error gracefully, or just let the entire thing
+      # crash?
```

Comment:
I would go with a try/catch. If you use hooks, I think they are automatically try/catch-ed for you.

Reply:
I think this method specifically will not use hooks, but when we add training loop fusion then it will be necessary to use hooks and it will be necessary for this to rescue gracefully in some way